fem/bilinearform.cpp

```diff
@@ -521,17 +531,26 @@ void BilinearForm::Assemble(int skip_zeros)
             bbfi[k]->AssembleElementMatrix(be, *eltrans, elemmat);
             elmat += elemmat;
          }
+         if (doftrans)
```

Comment:
I wanted to suggest a version (overload) of `TransformDual` (etc.) that only takes one argument that is modified in-place. Using it would simplify a lot of places like this, i.e., if you don't have `doftrans`, you do nothing, and don't need a temporary. Do you think that would be possible?

Reply:
That does sound like a good idea. I'll give it a try.

Reply:
The `DofTransformation` class was written to support rectangular transformations but I cannot remember exactly why. I have a vague recollection of discussing variable order with @tzanio in this context. An object like this could be used to reconstruct equivalent high-order DoFs from a set of low order DoFs. However, such an object would be setup and called in a different way not to mention that I think we have better ways to do such things now.

Unless @tzanio can think of a reason not to do this I'll rewrite the classes to assume square transformations with in-place arguments.

Reply:
I agree. Lets keep `DofTransformation` just for the square transformation case.

Reply:
I implemented this suggestion in PR #2386 .

If everyone's ok with it I can merge it into this PR for final review.

Reply:
> I implemented this suggestion in PR #2386 .
> 
> If everyone's ok with it I can merge it into this PR for final review.

I am ok with that. I'll need some more time to review this again. Thanks